### PR TITLE
Add logconvert script

### PR DESCRIPTION
**Pull request by simontegelid**
Make use of the log readers and writers to just convert between different log file formats.

### CODE REVIEW

**Comment by simontegelid** [comment]
I consider it done. Would be great with a few more tests though, I just tested it to the same extent as the other scripts.

**Comment by felixdivo** [comment]
Hey @simontegelid! This looks like a very nice addition to the library. Just a little question: Could we make the script accept the output as a second argument without specifying `-o`? For example: `logconvert.py input.file output.file`? I think that would be simpler.

File: can/logconvert.py
Comment (felixdivo) [inline can/logconvert.py:22]:
```suggestion
        help="Output filename, type dependent on suffix see can.LogReader.",
```

File: can/logconvert.py
Comment (felixdivo) [inline can/logconvert.py:32]:
```suggestion
        help="Maximum file size in bytes. Rotate log file when size threshold is reached.",
```

File: can/logconvert.py
Comment (felixdivo) [inline can/logconvert.py:14]: I don't think that we need this, do we?

```suggestion
```

File: can/logconvert.py
Comment (felixdivo) [inline can/logconvert.py:50]: We should probably close the file descriptors afterwards. `with LogReader(results.infile) as reader:` and similarly for the logger should be enough.

File: can/logconvert.py
Comment (felixdivo) [inline can/logconvert.py:63]:
```suggestion
        print("Interrupted! Quitting.")
```

**Review by felixdivo — CHANGES_REQUESTED** [review]

File: can/logconvert.py
Comment (simontegelid) [reply can/logconvert.py:63]: I agree that just "pass" isn't the best idea here. I think it would be better with sys.exit(1) though, instead of writing something to stdout and then exit with 0. What do you think?

File: can/logconvert.py
Comment (simontegelid) [reply can/logconvert.py:50]: Yes, will fix this.

File: can/logconvert.py
Comment (simontegelid) [reply can/logconvert.py:14]: No, not really. But I added it to keep consistency with the other scripts. I think it should be here until all the scripts are cleaned up.

**Comment by simontegelid** [comment]
> Hey @simontegelid! This looks like a very nice addition to the library. Just a little question: Could we make the script accept the output as a second argument without specifying `-o`? For example: `logconvert.py input.file output.file`? I think that would be simpler.

I think the -o flag should be there to avoid mixing positional arguments of various types to avoid issues with future updates (eg. adding support for multiple input files or something)

File: can/logconvert.py
Comment (felixdivo) [reply can/logconvert.py:14]: 👍

I'll remove that line in a separate PR real quick so you can just apply this suggestion.

Comment (felixdivo) [reply can/logconvert.py:14]: See #1099

File: can/logconvert.py
Comment (felixdivo) [reply can/logconvert.py]: Sounds even better!

**Comment by felixdivo** [comment]
> I think the -o flag should be there to avoid mixing positional arguments of various types to avoid issues with future updates

Well, one could use flags like `-o additional.file`  then too (if such features ever gets added). I'd argue that the the 1-to-1 conversion will probably stay the most used variant and should thus be the most simple to use.

**Comment by simontegelid** [comment]
> > I think the -o flag should be there to avoid mixing positional arguments of various types to avoid issues with future updates
> 
> Well, one could use flags like `-o additional.file` then too (if such features ever gets added). I'd argue that the the 1-to-1 conversion will probably stay the most used variant and should thus be the most simple to use.

Fair enough. Fixed now!

File: can/logconvert.py
Comment (felixdivo) [inline can/logconvert.py:37]:
```suggestion
        help="Input filename. Type dependent on suffix, see can.LogReader.",
```

File: can/logconvert.py
Comment (felixdivo) [inline can/logconvert.py:44]:
```suggestion
        help="Output filename. The type is dependent on the suffix, see can.Logger.",
```

File: can/logconvert.py
Comment (felixdivo) [reply can/logconvert.py]: Seems like this is the only thing left to be done before merging.

Comment (simontegelid) [reply can/logconvert.py]: I solved this with a .stop() in the finally clause to avoid using contextmanagers or so due to the SizedRotatingLogger vs Logger instances. Will do the same thing as using `with`. Is that good enough?

File: can/logconvert.py
Comment (felixdivo) [inline can/logconvert.py:64]: I think this should do, it's a couple of lines shorter and maybe a little bit easier to understand:

```suggestion
    with LogReader(args.input) as reader:

        if args.file_size:
            logger = SizedRotatingLogger(
                base_filename=args.output, max_bytes=args.file_size
            )
        else:
            logger = Logger(filename=args.output)

        with logger:
            try:
                for m in reader:  # pylint: disable=not-an-iterable
                    logger(m)
            except KeyboardInterrupt:
                sys.exit(1)
```

File: can/logconvert.py
Comment (felixdivo) [reply can/logconvert.py]: I commented below. I think you can just apply it as is and we can merge it. Functionally, they should do pretty much the same.

File: can/logconvert.py
Comment (simontegelid) [reply can/logconvert.py:64]: Just tested it, but `SizedRotatingLogger` doesn't implement `__enter__` so `python -m can.logconvert -s 1000 log.log log.asc` doesn't work.

Comment (felixdivo) [reply can/logconvert.py:64]: Whoops! That's a bug, I suppose. `SizedRotatingLogger` should inherit from `Logger`, which would then work since `BaseIOHandler` implements the context manager protocol.

**Comment by simontegelid** [comment]
So, I rebased this whole thing, therefore the force-push. Now it works as intended, using `with logger`, etc.

**Review by felixdivo — APPROVED** [review]

**Comment by felixdivo** [comment]
@hardbyte Do you approve?

**Comment by hardbyte** [comment]
@felixdivo Indeed I do, sorry to keep you waiting for my review on so many PRs. Where you are confident of a change I'm happy for you to decide to merge without my review. 

@simontegelid thanks for the feature!

**Comment by felixdivo** [comment]
@hardbyte Thanks for having a look! Your offer sounds good, I think if we have two or more reviews, it's already quite a good assurance that a change is valid and wanted. 😃

**Pull request merged**
